src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -656,70 +653,204 @@ IEnumerator IEnumerable.GetEnumerator()
+                    {
+                        return true;
+                    }
+
+                    key = subKey;
+                }
+                else
                 {
-                    continue;
+                    return false;
                 }
+            }
 
-                if (!key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
+            // Everything is prefixed by the empty string
+            if (prefix.Length == 0)
+            {
+                return true;
+            }
+            else
+            {
+                var charAfterPrefix = key[prefix.Length];
+                switch (charAfterPrefix)
                 {
+                    case '[':
+                    case '.':
+                        return true;
+                }
+            }
 
-                    if (key.StartsWith("[", StringComparison.OrdinalIgnoreCase))
-                    {
-                        var subKey = key.Substring(key.IndexOf('.') + 1);
+            return false;
+        }
 
-                        if (!subKey.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
-                        {
-                            continue;
-                        }
+        public PrefixEnumerable FindKeysWithPrefix(string prefix)
+        {
+            if (prefix == null)
+            {
+                throw new ArgumentNullException(nameof(prefix));
+            }
 
-                        if (string.Equals(prefix, subKey, StringComparison.Ordinal))
-                        {
-                            yield return entry;
-                            continue;
-                        }
+            return new PrefixEnumerable(this, prefix);
+        }
 
-                        key = subKey;
-                    }
-                    else
-                    {
-                        continue;
-                    }
+        public struct PrefixEnumerable : IEnumerable<KeyValuePair<string, ModelState>>
+        {
+            private readonly ModelStateDictionary _dictionary;
+            private readonly string _prefix;
+
+            public PrefixEnumerable(ModelStateDictionary dictionary, string prefix)
+            {
+                if (dictionary == null)
+                {
+                    throw new ArgumentNullException(nameof(dictionary));
                 }
 
-                // Everything is prefixed by the empty string
-                if (prefix.Length == 0)
+                if (prefix == null)
                 {
-                    yield return entry;
+                    throw new ArgumentNullException(nameof(prefix));
                 }
-                else
+
+                _dictionary = dictionary;
+                _prefix = prefix;
+            }
+
+            public PrefixEnumerator GetEnumerator()
+            {
+                return _dictionary == null ? new PrefixEnumerator() : new PrefixEnumerator(_dictionary, _prefix);
```

Comment:
Well, it's an inner `struct` that's `public` only because it has to be and we don't use it that way but :ok: